File: java/MinDeque.java
// data structure that allows insertion and deletion from both ends 
// also provides minimum element at any given time

public class MinDeque {
  private Node head;
  private Node tail;
  private Node minNode;      // node pointing to min element

  private class Node {
    int val;
    Node prev, next;
    Node(int val) {
      this.val = val;
    }
  }
  public void pushFront(int val) {
    Node newNode = new Node(val);
    if(head == null) {
      head = tail = minNode = newNode;
    } else {
      newNode.next = head;
      head.prev = newNode;
      head = newNode;
      minNode = min(minNode, newNode);
    }
  }
  public void pushBack(int val) {
    Node newNode = new Node(val);
    if(tail == null) {
      head = tail = minNode = newNode;
    } else {
      tail.next = newNode;
      newNode.prev = tail;
      tail = newNode;
      minNode = min(minNode, newNode);
    }
  }
  public int popFront() {
    if(head == null) {
      throw new IllegalStateException("deque is empty");
    }
    int val = head.val;
    head = head.next;
    if(head != null) {
      head.prev = null;
      minNode = min(minNode, head);
    } else {
      tail = null;
      minNode = null;
    }
    return val;
  }
  public int popBack() {
    if(tail == null) {
      throw new IllegalStateException("deque is empty");
    }
    int val = tail.val;
    tail = tail.prev;
    if(tail != null) {
      tail.next = null;
      minNode = min(minNode, tail);
    } else {
      head = null;
      minNode = null;
    }
    return val;
  }
  public int getMin() {
    if(minNode == null) {
      throw new IllegalStateException("deque is empty");
    }
    return minNode.val;
  }
  private Node min(Node a, Node b) {
    return a.val <= b.val ? a : b;
  }
}

File: python/min_deque.py
# python implementation using double ended queue from collections module and a min heap 

from collections import deque
import heapq

class MinDeque:
  def __init__(self):
    self.deque = deque()
    self.min_heap = []

  def push_front(self, val):
    self.deque.appendleft(val)
    heapq.heappush(self.min_heap, val)

  def push_back(self, val):
    self.deque.append(val)
    heapq.heappush(self.min_heap, val)

  def pop_front(self):
    if not self.deque:
      raise IndexError("deque is empty")
    val = self.deque.popleft()
    self.min_heap.remove(val)
    heapq.heapify(self.min_heap)
    return val 

  def pop_back(self):
    if not self.deque:
      raise IndexError("deque is empty")
    val = self.deque.pop()
    self.min_heap.remove(val)
    heapq.heapify(self.min_heap)
    return val 

  def get_min(self):
    if not self.min_heap:
      raise IndexError("deque is empty")
    return self.min_heap[0]
